### PR TITLE
Basic model_evaluation tests

### PR DESCRIPTION
**Pull request by jack-fs**
Tests:
- Bootstrapping mean recovers mean evaluation with high probability
- Cross validation and bootstrapping are deterministic, given fixed random seed
- Fuzz testing of bootstrap_model, cross_validate model
    - Note that **fuzz-testing search space has been reduced so that current implementation passes.** 
        - TODO comments in test_model_evaluation show how to re-introduce regions of search space that yield poor behaviour.
    - ValueErrors are reduced to warnings in fuzz-testing. These are almost all due to e.g. overflows (to be expected given pathological test data).

### CODE REVIEW

File: cinspect/model_evaluation.py [hunk 1]
```diff
@@ -116,8 +117,13 @@ def bootstrap_model(
         start = time.time()
         Xb, yb, indicesb = resample(X, y, indices)
 
-        if groups:
-            estimator.fit(Xb, yb, groups=indicesb)
+        if groups :
```
Comment (dsteinberg) [inline cinspect/model_evaluation.py:120]: Nice catch!

File: tests/test_model_evaluation.py [hunk 2]
```diff
@@ -72,3 +85,305 @@ def test_eval_function_calls(eval_func):
...
+        """Return (random) evaluation."""
+        return self._results
+
+
+model_evaluators = [crossval_model, bootstrap_model]
+random_seeds = [42, np.random.RandomState()]
+
+
+@pytest.mark.parametrize(
+    "eval_func, random_state", itertools.product(model_evaluators, random_seeds)
```
Comment (dsteinberg) [inline tests/test_model_evaluation.py:128]: I don't think you actually need `product` here - I believe the default behaviour of `parametrize` is to do a Cartesian product of the arguments. See how here:

https://github.com/gradientinstitute/aboleth/blob/master/tests/test_kernels.py#L28

File: tests/test_model_evaluation.py [hunk 3]
```diff
@@ -72,3 +85,305 @@ def test_eval_function_calls(eval_func):
...
+        [evaluator],
+        replications=n_bootstrap_replications,
+        random_state=random_state,
+    )
+    results = evaluator_.get_results()
+    bs_mean = np.mean(results)
+
+    # with 95% probability,
+    # mean in bs_mean +- bound.
+    bound = 1.96 * (stdev / (np.sqrt(n_bootstrap_replications)))
```
Comment (dsteinberg) [inline tests/test_model_evaluation.py:241]: This is a constant - so you probably don't need to recompute it every call. You could potentially make these module "settings" variables, e.g.

```python
N_BOOTSTRAP_REPLICATIONS = 30
BS_MEAN = 0
...

def _test_bootstrap_samples_from_eval_distribution(random_state):
 ...
```

**Review by dsteinberg — APPROVED** [review]
Very slick!!

A few little comments - up to you whether you want to change anything.

Otherwise LGTM!

File: tests/test_model_evaluation.py (hunk 3 above)
Comment (jack-fs) [reply tests/test_model_evaluation.py:241]: I think I'll leave this where it is for now, more or less. Agreed that having test settings/a config would be useful. I'll probably end up splitting this test up into multiple sub-components that I can re-use for e.g. numerical tests (e.g. "run this test n times and fail if it fails more than once"); I'll probably end up parametrising this particular function over the evaluation distribution, so I'll leave it for the next round of tests for now.

**Pull request merged**